src/puppetlabs/puppetdb/command/dlo.clj

```diff
@@ -1,210 +1,182 @@
+  (let [tmp (Files/createTempFile dir
+                                  (str "tmp-err-" id \- command) ""
+                                  (make-array FileAttribute 0))]
+    ;; Leave the temp file if something goes wrong.
+    (with-open [out (java.io.PrintWriter. (.toFile tmp))]
+      (write-failure-metadata out attempts))
+    (let [dest (.resolve dir (err-filename id metadata))
+          moved? (try
+                   (Files/move tmp dest (copts [copt-atomic]))
+                   (catch FileAlreadyExistsException ex
+                     true)
+                   (catch UnsupportedOperationException ex
+                     false)
+                   (catch AtomicMoveNotSupportedException ex
+                     false))]
+      (when-not moved?
+        (Files/move tmp dest (copts [copt-replace])))
+      dest)))
+
+
+;;; Public interface
+
+(defn initialize
+  "Initializes the dead letter office (DLO), at the given path (a Path),
+   creating the directory if it doesn't exist, adds related metrics to
+   the registry, and then returns a representation of the DLO."
+  [path registry]
+  (try
+    (Files/createDirectory path (make-array FileAttribute 0))
+    (catch FileAlreadyExistsException ex
+      (when-not (Files/isDirectory path (make-array LinkOption 0))
+        (throw (Exception. (trs "DLO path {0} is not a directory" path))))))
+  {:path path
+   :registry registry
+   :metrics (atom (metrics-for-dir registry path))})
+
+(defn discard-cmdref
+  "Stores information about the failed `stockpile` `cmdref` to the
+  `dlo` (a Path) for later inspection.  Saves two files named like
+  this:
+    10291-1469469689-cat-4-foo.org.json
+    10291-1469469689-cat-4-foo.org-err.txt
+  The first contains the failed command itself, and the second details
+  the cause of the failure.  The command will be moved from the
+  `stockpile` queue to the `dlo` directory (a Path) via
+  stockpile/discard.  Returns {:info Path :command Path}."
+  [cmdref stockpile dlo]
+  (let [{:keys [path registry metrics]} dlo
+        {:keys [command received attempts]} cmdref
+        entry (cmdref->entry cmdref)
+        cmd-dest (.resolve path (entry-cmd-data-filename entry))]
+    ;; We're going to assume that our moves will be atomic, and if
+    ;; they're not, that we don't care about the possibility of
+    ;; partial dlo messages.  If needed, the existence of the err file
+    ;; can be used as an indicator that the dlo message is complete.
+    (stock/discard stockpile entry cmd-dest)
+    (let [id (stock/entry-id entry)
+          metadata (stock/entry-meta entry)
+          info-dest (store-failed-command-info id metadata command
+                                               attempts
+                                               path)]
+      (swap! metrics ensure-cmd-metrics registry command)
+      (update-metrics @metrics command (Files/size cmd-dest))
+      {:info info-dest
+       :command cmd-dest})))
+
+(defn discard-bytes
```

Comment:
The only real purpose of this function is to discard a command when we don't know much about it. Specifically this happens when migrating an ActiveMQ message into stockpile and some error occurs. If it's an old message and we can't parse the payload, we won't know the command type, certname, etc. This allows us to put that message somewhere that the user can follow up on